src/Functions/FunctionHelpers.cpp

```diff
@@ -120,73 +119,57 @@ void validateArgumentsImpl(const IFunction & func,
+                            descriptor.name,
                             func.getName(),
-                            (descriptor.expected_type_description ? String(", expected ") + descriptor.expected_type_description : String{}),
-                            (arg.type ? ", got " + arg.type->getName() : String{}));
+                            descriptor.type_name,
+                            arg.type ? arg.type->getName() : "<?>");
     }
 }
 
 }
 
 int FunctionArgumentDescriptor::isValid(const DataTypePtr & data_type, const ColumnPtr & column) const
 {
-    if (type_validator_func && (data_type == nullptr || !type_validator_func(*data_type)))
+    if (name.empty() || type_name.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "name or type_name are not set");
+
+    if (type_validator && (data_type == nullptr || !type_validator(*data_type)))
         return ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT;
 
-    if (column_validator_func && (column == nullptr || !column_validator_func(*column)))
+    if (column_validator && (column == nullptr || !column_validator(*column)))
         return ErrorCodes::ILLEGAL_COLUMN;
 
     return 0;
 }
 
-void validateFunctionArgumentTypes(const IFunction & func,
-                                   const ColumnsWithTypeAndName & arguments,
-                                   const FunctionArgumentDescriptors & mandatory_args,
-                                   const FunctionArgumentDescriptors & optional_args)
+void validateFunctionArguments(const IFunction & func,
+                               const ColumnsWithTypeAndName & arguments,
+                               const FunctionArgumentDescriptors & mandatory_args,
+                               const FunctionArgumentDescriptors & optional_args)
 {
     if (arguments.size() < mandatory_args.size() || arguments.size() > mandatory_args.size() + optional_args.size())
     {
-        auto join_argument_types = [](const auto & args, const String sep = ", ")
-        {
-            String result;
-            for (const auto & a : args)
-            {
-                using A = std::decay_t<decltype(a)>;
-                if constexpr (std::is_same_v<A, FunctionArgumentDescriptor>)
-                {
-                    if (a.argument_name)
-                        result += "'" + std::string(a.argument_name) + "' : ";
-                    if (a.expected_type_description)
-                        result += a.expected_type_description;
-                }
-                else if constexpr (std::is_same_v<A, ColumnWithTypeAndName>)
-                    result += a.type->getName();
-
-                result += sep;
-            }
-
-            if (!args.empty())
-                result.erase(result.end() - sep.length(), result.end());
-
-            return result;
-        };
+        auto argument_singular_or_plural = [](const auto & args){ return fmt::format("argument{}", args.size() != 1 ? "s" : ""); };
```

Comment:
```suggestion
        auto argument_singular_or_plural = [](const auto & args) -> std::string_view { return args.size() != 1 ? "arguments" : "argument"); };
```

it's only used in `fmt::format` so no need to allocate and create a string each time, we can use constants for that.

Reply:
Makes sense, done.